base/intfuncs.jl

```diff
@@ -490,14 +490,19 @@ julia> prevpow(4, 16)
 16
 ```
 """
-function prevpow(a::Real, x::Real)
+function prevpow(a::Real, x::T) where T <: Real
     x < 1 && throw(DomainError(x, "`x` must be ≥ 1."))
     # See comment in nextpos() for a == special case.
     a == 2 && isa(x, Integer) && return _prevpow2(x)
     a <= 1 && throw(DomainError(a, "`a` must be greater than 1."))
     n = floor(Integer,log(a, x))
-    p = a^(n+1)
-    p <= x ? p : a^n
+    p = T(a)^n
+    if a isa Integer && x isa Integer
+        wp, overflow = mul_with_overflow(T(a), x)
```

Comment:
```suggestion
    if p isa Integer && x isa Integer
        wp, overflow = mul_with_overflow(p, oftype(p, x))
```

Reply:
This is wrong. `wp` needs to be of the same type as `x` to prevent spurious overflow.

Reply:
In that case maybe restrict the signature to `prevpow(a::T, x::T) where {T}` and use `prevpow(promote(a, x)...)` for the fallback definition.

Reply:
I think that's probably the best answer.

Reply:
wait, that doesn't work. The desired behavior is to follow the eltype of the 2nd argument, not the promotion.

Reply:
nevermind, it's actually the first argument. `prevpow` is a weird function.